packages/kotti-ui/source/props.ts

```diff
@@ -0,0 +1,198 @@
+import { PropOptions, PropType } from '@vue/composition-api'
+import { cloneDeep } from 'lodash'
+import { z } from 'zod'
+
+import { MapValueType } from './types/utilities'
+
+const DEBUG_MAKE_PROPS = false as const // enable to print debug log
+const DEBUG_WALK_SHAPE = false as const // enable to print debug log
+
+const propValidator = <SCHEMA extends z.ZodTypeAny>(
+	schema: SCHEMA,
+	key: string,
+) => {
+	const validator = (value: unknown) => {
+		const result = schema.safeParse(value)
+
+		if (result.success) return true
+
+		/* eslint-disable no-console */
+		console.group(`propValidator found issues with prop “${key}”`)
+
+		// HACK: 'error' in result is necessary as `ts-jest` doesn’t see that result.success was already properly checked to be falsy and that error now exists
+		if ('error' in result) console.error(result.error)
+
+		if (Array.isArray(value)) console.table(cloneDeep(value))
+		else console.log(cloneDeep(value))
+
+		console.trace()
+
+		console.groupEnd()
+		/* eslint-enable no-console */
+
+		return false
+	}
+
+	// assign name for better debugging and to improve props rendering in the documentation
+	Object.defineProperty(validator, 'name', {
+		value: `propValidator('${key}')`,
+		writable: false,
+	})
+
+	return validator
+}
+
+// eslint-disable-next-line sonarjs/cognitive-complexity
+const walkShape = <SHAPE extends z.ZodTypeAny>(shape: SHAPE): Set<string> => {
+	/* eslint-disable no-console */
+	const types = new Set<string>()
+
+	let shapePointer = shape
+
+	while (shapePointer) {
+		const { typeName } = shapePointer._def
+
+		if (DEBUG_WALK_SHAPE)
+			console.log(
+				`walkTypes: found typeName “${typeName}” (${typeof typeName})`,
+			)
+
+		if (typeName === 'ZodUnion') {
+			if (DEBUG_WALK_SHAPE) console.group('walkTypes: walking union')
+
+			const options = shapePointer._def.options as Array<z.ZodTypeAny>
+
+			// walk every option in the enum and merge result sets into types
+			for (const typeSet of options.map((x) => walkShape(x)))
+				for (const type of typeSet) types.add(type)
+
+			if (DEBUG_WALK_SHAPE) console.groupEnd()
+		} else {
+			if (DEBUG_WALK_SHAPE) console.log(`walkTypes: pushing “${typeName}”`)
+			types.add(typeName)
+		}
+
+		shapePointer = shapePointer._def.innerType
+	}
+
+	return types
+	/* eslint-enable no-console */
+}
+
+/**
+ * These types do not influence the generation of Vue’s `propName.type` Arrays and are therefore discarded
+ */
+const ignoredZodTypes = new Set(['ZodDefault', 'ZodNullable', 'ZodOptional'])
+
+/**
+ * This maps the internal zod name of a type to the the constructor that Vue expects in `propName.type`
+ */
+const zodToVueTypes = new Map<
+	string,
+	Exclude<PropType<unknown>, Array<unknown>>
```

Comment:
if they expose it, I personally think it might be better to use `PropConstructor`

Vue's definition of `PropType<T>` is `PropConstructor | PropConstructor[]`
and you don't want the array of whatever PropConstructor is 

Assumptions:
- they actually expose it
-  TS doesn't scream about it
-  you find it more readable

Reply:
Otherwise don't change

Reply:
Not exported:
Personally make a separate type here:
`type PropConstructor = Exclude<PropType<unknown>, Array<unknown>>`

with a comment explaining, that we would've otherwise used the declared type from `Vue` if it were exported